### PR TITLE
C++ operator new/delete bindings

### PR DESCRIPTION
**Pull request by TurkeyMan**
`std::allocator` depends on this.

### CODE REVIEW

**Comment by TurkeyMan** [comment]
I don't know what to do about that OSX fail. @jacob-carlborg?
Is it spurious? I'm not even sure what it wants from me...

This PR is holding all my work up, so I'd like to get it through.

**Comment by jacob-carlborg** [comment]
> I don't know what to do about that OSX fail. @jacob-carlborg?
Is it spurious? I'm not even sure what it wants from me...

Seems to be the same problem as the other PR. I don't think it's something wrong with your PR, I think it's the machine the CI is running on. Try removing `-g` from `CFLAGS` here:

https://github.com/dlang/druntime/blob/227a925a1892deaab7a64a1bda576864d04dd528/test/common.mak#L26

But since it's a shared makefile it might cause other tests to fail. Alternatively don't use the `CXXFLAGS` variable here to avoid `-g` being added:

https://github.com/dlang/druntime/blob/227a925a1892deaab7a64a1bda576864d04dd528/test/stdcpp/Makefile

**Comment by thewilsonator** [comment]
Can confirm, removal of `-g` now results in 
```
c++ -m64 -fPIC -Wall -std=c++11 -stdlib=libc++ -o generated/osx/debug/64/array src/array.cpp ./generated/osx/debug/64/array_d.o ~/d/druntime/generated/osx/debug/64/libdruntime.a -lpthread
ld: warning: could not create compact unwind for __D2rt9backtrace5dwarf23traceHandlerOpApplyImplFxAPvMDFKmKxAaZiZi: offset of saved registers too far to encode
```
and the execrable run successfully with no output.

**Comment by thewilsonator** [comment]
```diff
diff --git a/test/common.mak b/test/common.mak
index fbd72ae..617375d 100644
--- a/test/common.mak
+++ b/test/common.mak
@@ -17,18 +17,20 @@ ROOT:=$(GENERATED)/$(OS)/$(BUILD)/$(MODEL)
 ifneq (default,$(MODEL))
 	MODEL_FLAG:=-m$(MODEL)
 endif
-CFLAGS:=$(MODEL_FLAG) $(PIC) -Wall
+CFLAGS_BASE:= $(MODEL_FLAG) $(PIC) -Wall
 DFLAGS:=$(MODEL_FLAG) $(PIC) -w -I../../src -I../../import -I$(SRC) -defaultlib= -debuglib= -dip1000
 # LINK_SHARED may be set by importing makefile
 DFLAGS+=$(if $(LINK_SHARED),-L$(DRUNTIMESO),-L$(DRUNTIME))
 ifeq ($(BUILD),debug)
 	DFLAGS += -g -debug
-	CFLAGS += -g
+    CFLAGS := $(CFLAGS_BASE) -g
 else
 	DFLAGS += -O -release
-	CFLAGS += -O3
+	CFLAGS := $(CFLAGS_BASE) -O3
 endif
+CXXFLAGS_BASE := $(CFLAGS_BASE) -std=c++11
 CXXFLAGS:=$(CFLAGS) -std=c++11
 ifeq (osx,$(OS))
 	CXXFLAGS+=-stdlib=libc++
+    CXXFLAGS_BASE+=-stdlib=libc++
 endif
diff --git a/test/stdcpp/Makefile b/test/stdcpp/Makefile
index 4497e45..1677a51 100644
--- a/test/stdcpp/Makefile
+++ b/test/stdcpp/Makefile
@@ -22,7 +22,7 @@ $(ROOT)/%.done : $(ROOT)/%
 $(ROOT)/%: $(SRC)/%.cpp $(SRC)/%_test.d
 	mkdir -p $(dir $@)
 	$(QUIET)$(DMD) $(DFLAGS) -main -unittest -c -of=$(ROOT)/$*_d.o $(SRC)/$*_test.d
-	$(QUIET)$(CXX) $(CXXFLAGS) -o $@ $< $(ROOT)/$*_d.o $(DRUNTIME) -lpthread
+	$(QUIET)$(CXX) $(CXXFLAGS_BASE) -o $@ $< $(ROOT)/$*_d.o $(DRUNTIME) -lpthread

 clean:
 	rm -rf $(GENERATED)
```
Should do the job.

**Comment by TurkeyMan** [comment]
I mean... yeah.
It feels kinda brutal to hack the makefiles that way; as if they weren't already complicated enough.

**Comment by TurkeyMan** [comment]
Well, I pushed it... are we happy to commit that patch?

**Comment by thewilsonator** [comment]
> are we happy to commit that patch?

Yes. I'd rather you not get stuck on things that, while nice, are not necessary for moving things forward. We can fix these things up later.

**Comment by TurkeyMan** [comment]
And #2444, which is basically this, just wrapped in `std::allocator` API.

**Comment by thewilsonator** [comment]
I know, I know. I can can only go as fast as the CI pipeline ;)

File: src/core/stdcpp/new_.d [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ * Authors:   Manu Evans
+ * Source:    $(DRUNTIMESRC core/stdcpp/new_.d)
+ */
+
+module core.stdcpp.new_;
+
+import core.stdcpp.xutility : __cpp_sized_deallocation, __cpp_aligned_new;
+
+// TODO: this really should come from __traits(getTargetInfo, "defaultNewAlignment")
+enum size_t __STDCPP_DEFAULT_NEW_ALIGNMENT__ = 16;
```
Comment (jacob-carlborg) [inline src/core/stdcpp/new_.d]: Is this necessary to be public?

File: src/core/stdcpp/xutility.d [hunk 2]
```diff
@@ -11,6 +11,22 @@
...
+    cpp11 = 201103,
+    cpp14 = 201402,
+    cpp17 = 201703
+}
+
+enum __cplusplus = __traits(getTargetInfo, "cppStd");
+
+// wrangle C++ features
+enum __cpp_sized_deallocation = __cplusplus >= CppStdRevision.cpp14 ? 201309 : 0;
+enum __cpp_aligned_new = __cplusplus >= CppStdRevision.cpp17 ? 201606 : 0;
```
Comment (jacob-carlborg) [inline src/core/stdcpp/xutility.d]: Is it necessary to all these symbols are public?

**Comment by TurkeyMan** [comment]
I dunno... They are in C++...

File: src/core/stdcpp/new_.d (hunk 1 above)
Comment (TurkeyMan) [reply src/core/stdcpp/new_.d]: Is it conceivable that someone binds their own containers that respond to this value?
It does need to be a `getTargetInfo` though; this value is wrong.

File: src/core/stdcpp/xutility.d (hunk 2 above)
Comment (TurkeyMan) [reply src/core/stdcpp/xutility.d]: I dunno... They are in C++...?

File: src/core/stdcpp/new_.d (hunk 1 above)
Comment (jacob-carlborg) [reply src/core/stdcpp/new_.d]: I don't know enough about this but it looks like implementation details. BTW, they miss Ddoc comments.

Comment (TurkeyMan) [reply src/core/stdcpp/new_.d]: Yeah, I didn't put ddoc because I wasn't sure if public or not.

Comment (jacob-carlborg) [reply src/core/stdcpp/new_.d]: Either have them as public and documented or not public.

**Pull request merged**